### PR TITLE
Fix MAGN 9631 Large value range expression crashes on committing the CBN

### PR DESCRIPTION
**Pull request by ke-yu**
### Purpose

Fix [MAGN 9631 Large value range expression crashes on committing the CBN] (http://adsk-oss.myjetbrains.com/youtrack/issue/MAGN-9631)

There are two kinds of exceptions may be thrown out in range expression generation. 

First is `OutOfMemory` crash which is thrown out when creating a large `StackValue` array. CLR puts a hard limit on the size of object (<= 2GB), no matter the application is 32-bit or 64-bit (https://msdn.microsoft.com/en-us/library/ms241064(VS.80).aspx). The DS runtime needs to catch this kind of exception and log a runtime warning.

Second is integer overflow exception. It may be thrown out when the number of item in range expression is over the limit of 32 bit integer. For example, in expression `1..10000000000`.  Adding checking to avoid this exception.

untitled

### Declarations

Check these if you believe they are true

- [x] The code base is in a better state after this PR
- [ ] Is documented according to the [standards](https://github.com/DynamoDS/Dynamo/wiki/Coding-Standards)
- [ ] The level of testing this PR includes is appropriate
- [x] User facing strings, if any, are extracted into `*.resx` files
- [ ] All tests pass using the self-service CI.
- [ ] Snapshot of UI changes, if any.

### Reviewers
@Benglin

### CODE REVIEW

File: src/Engine/ProtoCore/Lang/BuiltInFunctionEndPoint.cs [hunk 1]
```diff
@@ -1158,31 +1100,43 @@ internal static StackValue[] GenerateRangeByAmount(decimal start, long amount, d
...
-                       Resources.kInvalidAmountInRangeExpression);
+                    runtimeCore.RuntimeStatus.LogWarning(WarningID.kInvalidArguments, Resources.kInvalidAmountInRangeExpression);
                     return null;
                 }
-                decimal stepsize = new decimal(svStep.ToDouble().RawDoubleValue);
-                return GenerateRangeByAmount(start, amount, stepsize, isIntRange);
+
+                if (amount == 0)
+                {
+                    return new StackValue[] { };
```
Comment (Benglin) [inline src/Engine/ProtoCore/Lang/BuiltInFunctionEndPoint.cs]: Other code paths returned `null` instead, is there a difference between `new StackValue[] { }` and `null` for the callers of this method?

File: src/Engine/ProtoCore/Lang/BuiltInFunctionEndPoint.cs [hunk 2]
```diff
@@ -1110,46 +1067,31 @@ internal static StackValue[] GenerateRangeByAmount(decimal start, long amount, d
...
-                                                         bool hasAmountOp,
-                                                         RuntimeCore runtimeCore)
+        private static StackValue[] GenerateNumericRange(
+            StackValue svStart,
+            StackValue svEnd,
+            StackValue svStep,
+            StackValue svOp,
+            bool hasStep,
+            bool hasAmountOp,
+            RuntimeCore runtimeCore)
```
Comment (Benglin) [inline src/Engine/ProtoCore/Lang/BuiltInFunctionEndPoint.cs]: I'm wondering if there are test cases to test this method with all combinations (various `if-else` code paths) of range expression. Do you think it is covered enough?

File: test/Engine/ProtoTest/TD/MultiLangTests/RangeExpressions.cs [hunk 3]
```diff
@@ -2134,5 +2134,41 @@ public void AlphabetSequenceNegativeTestCasesAssociative()
...
+        }
+
+        [Test]
+        [Category("Regression")]
+        public void TestRangeExpressionOverLimit04()
+        {
+            string src = @"x = 1..10..0.00000000001;";
+            thisTest.RunScriptSource(src);
+            thisTest.VerifyRuntimeWarningCount(1);
+        }
```
Comment (Benglin) [inline test/Engine/ProtoTest/TD/MultiLangTests/RangeExpressions.cs]: Thanks for adding the test cases!

**Comment by Benglin** [comment]
Few clarifications and we're good to go.

File: src/Engine/ProtoCore/Lang/BuiltInFunctionEndPoint.cs (hunk 1 above)
Comment (ke-yu) [inline src/Engine/ProtoCore/Lang/BuiltInFunctionEndPoint.cs]: empty list is a valid return return, `null` basically means there is an error in range expression, and normally the runtime will log a warning.

File: src/Engine/ProtoCore/Lang/BuiltInFunctionEndPoint.cs (hunk 2 above)
Comment (ke-yu) [inline src/Engine/ProtoCore/Lang/BuiltInFunctionEndPoint.cs]: I think so :-), we have over 70 test cases for range expression in associative language block and in imperative language blocks (`ProtoTest.TD.MultiLangTests.RangeExpressions`). And a lot of other test cases which more or less use range expressions.

**Pull request merged**